analysis/complex.lean

```diff
@@ -0,0 +1,342 @@
+  inv              := has_inv.inv,
+  mul_one          := by crunch,
+  one_mul          := by crunch,
+  mul_comm         := by crunch3,
+  mul_assoc        := by crunch,
+  left_distrib := by begin
+    intros,
+    apply eq_of_re_eq_and_im_eq,
+    { rw [mul_re,add_re,add_re,add_im,mul_re,mul_re],
+    simp [add_mul,mul_add] },
+    { rw [mul_im,add_re,add_im,add_im,mul_im,mul_im],
+    simp [add_mul,mul_add] },
+  end,
+--  left_distrib     := by crunch,
+  right_distrib    := by begin
+    intros,
+    apply eq_of_re_eq_and_im_eq,
+    { rw [mul_re,add_re,add_re,add_im,mul_re,mul_re],
+    simp [add_mul,mul_add] },
+    { rw [mul_im,add_re,add_im,add_im,mul_im,mul_im],
+    simp [add_mul,mul_add] },
+  end,
+  zero_ne_one      := begin
+    intro H,
+    suffices : ((0:real):complex).re = ((1:real):complex).re,
+    { revert this,
+      apply zero_ne_one },
+    { show (0:complex).re = (1:complex).re,
+    rw [←H],refl},
+    end,
+  mul_inv_cancel   := begin
+    intros z H,
+    have H2 : norm_squared z ≠ 0 := ne_of_gt (norm_squared_pos_of_nonzero z H),
+    apply eq_of_re_eq_and_im_eq,
+    { unfold has_inv.inv complex.inv,
+      rw [mul_re],
+      show z.re * (z.re / norm_squared z) -
+      z.im * (-z.im / norm_squared z) =
+    (1:complex).re,
+      suffices : z.re*(z.re/norm_squared z) + -z.im*(-z.im/norm_squared z) = 1,
+        by simpa,
+      rw [←mul_div_assoc,←mul_div_assoc,neg_mul_neg,div_add_div_same],
+      unfold norm_squared at *,
+      exact div_self H2},
+    { suffices : z.im * (z.re / norm_squared z) + z.re * (-z.im / norm_squared z) = 0,
+      by simpa,
+    rw [←mul_div_assoc,←mul_div_assoc,div_add_div_same],
+    simp [zero_div,mul_comm],
+    }
+  end,
+  inv_mul_cancel   := begin
+    intros z H,
+    have H2 : norm_squared z ≠ 0 := ne_of_gt (norm_squared_pos_of_nonzero z H),
+    apply eq_of_re_eq_and_im_eq,
+    { unfold has_inv.inv complex.inv,
+     rw [mul_re],
+     show (z.re / norm_squared z) * z.re -
+      (-z.im / norm_squared z) * z.im =
+    (1:complex).re,
+      suffices : z.re*(z.re/norm_squared z) + -z.im*(-z.im/norm_squared z) = 1,
+        by simpa [mul_comm],
+      rw [←mul_div_assoc,←mul_div_assoc,neg_mul_neg,div_add_div_same],
+      unfold norm_squared at *,
+      exact div_self H2 },
+    unfold has_inv.inv complex.inv,
+    rw [mul_im],
+    show (z.re / norm_squared z) * z.im +
+      (-z.im / norm_squared z) * z.re =
+    (1:complex).im,
+    suffices : z.im * (z.re / norm_squared z) + z.re * (-z.im / norm_squared z) = 0,
+      by simpa [mul_comm],
+    rw [←mul_div_assoc,←mul_div_assoc,div_add_div_same],
+    simp [zero_div,mul_comm]
+  end,
+  /-
+  inv_zero         := begin
+  unfold has_inv.inv complex.inv add_comm_group.zero,
+  apply eq_of_re_eq_and_im_eq,
+  split;simp [zero_div],
+  end,
+  -/
+--  has_decidable_eq := by apply_instance,
+  ..complex.add_comm_group,
+  }
+
+theorem im_eq_zero_of_complex_nat (n : ℕ) : (n:complex).im = 0 :=
+begin
+induction n with d Hd,
```

Comment:
indention

Reply:
Are there automatic linters for Lean? In python all those comments would be handled by a single run of `autopep8`.

Reply:
Sorry, I thought I'd caught them all. In the past I never indented after a begin. I had a system with no initial indent and no {}s which I have used extensively for months now and it's only through talking about this PR that I have come to understand that it is not quite the right system.

In general I don't see the point of this indentation after a begin -- it just means that all of my proofs are indented 2 spaces.

Reply:
If you write tactic proofs all your proofs are indented 2 spaces. It would be different for proof terms, where you have:
```lean
let x := ... in
have ..,
  from _,
have .., by ...,
show ..., ...
```